### PR TITLE
Bug 1836896 - Hide statusbar when fullscreen on Android R+ devices

### PR DESCRIPTION
**Pull request by jackyzy823**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1836896

### CODE REVIEW

**Review by t-p-white — CHANGES_REQUESTED** [review]
Nice solution, works well 👍 could you also update the tests in ActivityTest please? See example of testing with SDK [version here](https://github.com/mozilla-mobile/firefox-android/blob/e0f85f8a669b676847e890235d4bdfa7840d7f2c/android-components/components/browser/menu2/src/test/java/mozilla/components/browser/menu2/view/MenuViewTest.kt#L64)

**Comment by t-p-white** [comment]
bors try

**Comment by jackyzy823** [comment]
Not very sure the test sufficient or not.

**Comment by t-p-white** [comment]
@jackyzy823 thank you so much for adding the tests! Wdyt about updating & duplicating the existing tests for enter and exit with the Android version? e.g:

-  `check setAsImmersive sets the correct flags`
-  `check exitImmersiveMode sets the correct flags`

becomes:

-  `check setAsImmersive sets the correct flags for Android R and above`
-  `check setAsImmersive sets the correct flags for Android below R`
-  `check exitImmersiveMode sets the correct flags for Android R and above`
-  `check exitImmersiveMode sets the correct flags for Android below R`

**Comment by jackyzy823** [comment]
@t-p-white I tried to do this ,but due to i'm not familiar with mocking.

doing so
```kotlin
    @Test
    @Config(sdk = [Build.VERSION_CODES.R])
    fun `check setAsImmersive sets the correct flags  on devices with Android R and above`() {
        `when`(window.getInsetsController()).thenReturn(windowInsetsController)

        activity.setAsImmersive()

        verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
        verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
        verify(window.decorView, never()).setOnSystemUiVisibilityChangeListener(any())
        verify(window).setDecorFitsSystemWindows(false)
    }
```
will  cause
```kotlin
        verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
        verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
```
not work.


It says
```
Wanted but not invoked:
view.setSystemUiVisibility(4);
-> at android.view.View.setSystemUiVisibility(View.java)
Actually, there were zero interactions with this mock.
```


If i don't mock with ``` `when`(window.getInsetsController()).thenReturn(windowInsetsController)``` , it argues ```java.lang.NullPointerException: Cannot invoke android.view.WindowInsetsController.hide(int) because  this.mInsetsController  is null
        at androidx.core.view.WindowInsetsControllerCompat$Impl30.hide(WindowInsetsControllerCompat.java:664)
        at androidx.core.view.WindowInsetsControllerCompat.hide(WindowInsetsControllerCompat.java:188)
        at mozilla.components.support.ktx.android.view.ActivityKt.setAsImmersive(Activity.kt:38)```

It would be helpful if you could figure out how to make this correct.

**Comment by t-p-white** [comment]
@jackyzy823 thanks for looking into this, I see the same error. My solution is to leave the existing tests as-is (just updating the `Test` name to include the Android version part) and adding new tests for Android R - something similar too:

```
    @Test
    @Config(sdk = [Build.VERSION_CODES.R])
    fun `check setAsImmersive sets the correct flags on devices with Android R and above`() {
        `when`(window.insetsController).thenReturn(windowInsetsController)

        activity.setAsImmersive()

        verify(window).setDecorFitsSystemWindows(false)
        verify(windowInsetsController).hide(WindowInsetsCompat.Type.systemBars())
        verify(windowInsetsController).systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
    }
```

```
    @Test
    @Config(sdk = [Build.VERSION_CODES.R])
    fun `check exitImmersiveMode sets the correct flags on devices with Android R and above`() {
        `when`(window.insetsController).thenReturn(windowInsetsController)

        activity.exitImmersiveMode()

        verify(decorView).setOnApplyWindowInsetsListener(null)
        verify(window).setDecorFitsSystemWindows(true)
        verify(windowInsetsController).show(WindowInsetsCompat.Type.systemBars())
    }
```

Wdyt?

**Comment by jackyzy823** [comment]
@t-p-white thanks for advice.

File: android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt [hunk 1]
```diff
@@ -63,14 +69,26 @@ class ActivityTest {
     }
 
     @Test
-    fun `check setAsImmersive sets the correct flags`() {
+    fun `check setAsImmersive sets the correct flags  on devices with Android Q and below`() {
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt:72]:
```suggestion
    fun `check setAsImmersive sets the correct flags on devices with Android Q and below`() {
```

File: android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt [hunk 2]
```diff
@@ -63,14 +69,26 @@ class ActivityTest {
...
         activity.setAsImmersive()
 
         verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
         verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
         verify(window.decorView, never()).setOnSystemUiVisibilityChangeListener(any())
     }
 
+    @Test
+    @Config(sdk = [Build.VERSION_CODES.R])
+    fun `check setAsImmersive sets the correct flags  on devices with Android R and above`() {
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt:82]:
```suggestion
    fun `check setAsImmersive sets the correct flags on devices with Android R and above`() {
```

File: android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt [hunk 3]
```diff
@@ -112,13 +130,25 @@ class ActivityTest {
...
         activity.exitImmersiveMode()
 
         verify(decorView, times(2)).systemUiVisibility = View.SYSTEM_UI_FLAG_VISIBLE
         verify(decorView).setOnApplyWindowInsetsListener(null)
     }
 
+    @Test
+    @Config(sdk = [Build.VERSION_CODES.R])
+    fun `check exitImmersiveMode sets the correct flags on devices with Android R and above`() {
+        `when`(window.getInsetsController()).thenReturn(windowInsetsController)
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt:143]:
```suggestion
        `when`(window.insetsController).thenReturn(windowInsetsController)
```

File: android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt [hunk 4]
```diff
@@ -63,14 +69,26 @@ class ActivityTest {
...
 
         verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_FULLSCREEN
         verify(decorView).systemUiVisibility = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION
         verify(window.decorView, never()).setOnSystemUiVisibilityChangeListener(any())
     }
 
+    @Test
+    @Config(sdk = [Build.VERSION_CODES.R])
+    fun `check setAsImmersive sets the correct flags  on devices with Android R and above`() {
+        `when`(window.getInsetsController()).thenReturn(windowInsetsController)
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt:83]:
```suggestion
        `when`(window.insetsController).thenReturn(windowInsetsController)
```

File: android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt [hunk 5]
```diff
@@ -63,14 +69,26 @@ class ActivityTest {
...
+
+        activity.setAsImmersive()
+
+        verify(window).setDecorFitsSystemWindows(false)
+        verify(windowInsetsController).hide(WindowInsetsCompat.Type.systemBars())
+        verify(windowInsetsController).systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
```
Comment (t-p-white) [inline android-components/components/support/ktx/src/test/java/mozilla/components/support/ktx/android/view/ActivityTest.kt:89]: nit: my formatter is putting `WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE` on a newline

**Comment by t-p-white** [comment]
@jackyzy823 np, thanks for adding the tests 🙂 if you could make the minor suggested changes, I think we're good to land. Thanks again for working on this bug, I'm sure the fix will be well received!

**Comment by t-p-white** [comment]
bors try

**Comment by t-p-white** [comment]
bors try

**Review by t-p-white — APPROVED** [review]

**Comment by t-p-white** [comment]
bors try

**Comment by t-p-white** [comment]
bors try

**Comment by t-p-white** [comment]
bors try-

**Comment by t-p-white** [comment]
bors try

**Pull request merged**